### PR TITLE
8254073: Tokenizer improvements (revised)

### PR DESCRIPTION
**Pull request by JimLaskey**
This is a full revision of https://github.com/openjdk/jdk/pull/435 which contained two 'out by one' bugs and was reverted.

This revision contains the changes of that pull request plus:

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavadocTokenizer.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavadocTokenizer.java
index 39d9eadcf3a..b8425ad1ecb 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavadocTokenizer.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavadocTokenizer.java
@@ -306,8 +306,9 @@ public class JavadocTokenizer extends JavaTokenizer {
      *
      * Thus, to find the source position of any position, p, in the comment
      * string, find the index, i, of the pair whose string offset
-     * ({@code map[i + SB_OFFSET] }) is closest to but not greater than p. Then,
-     * {@code sourcePos(p) = map[i + POS_OFFSET] + (p - map[i + SB_OFFSET]) }.
+     * ({@code map[i * NOFFSETS + SB_OFFSET] }) is closest to but not greater
+     * than p. Then, {@code sourcePos(p) = map[i * NOFFSETS + POS_OFFSET] +
+     *                                (p - map[i * NOFFSETS + SB_OFFSET]) }.
      */
     static class OffsetMap {
         /**
@@ -426,7 +427,7 @@ public class JavadocTokenizer extends JavaTokenizer {
             int start = 0;
             int end = size / NOFFSETS;
 
-            while (start < end - NOFFSETS) {
+            while (start < end - 1) {
                 // find an index midway between start and end
                 int index = (start + end) / 2;
                 int indexScaled = index * NOFFSETS;
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java
index 2472632dbcd..7584b79044b 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java
@@ -221,48 +221,49 @@ public class UnicodeReader {
     private boolean unicodeEscape() {
         // Start of unicode escape (past backslash.)
         int start = position + width;
-        int index;
+
+        // Default to backslash result, unless proven otherwise.
+        character = '\\';
+        width = 1;
 
         // Skip multiple 'u'.
+        int index;
         for (index = start; index < length; index++) {
             if (buffer[index] != 'u') {
                 break;
             }
         }
 
-        // Needs to be at least backslash-u.
-        if (index != start) {
-            // If enough characters available.
-            if (index + 4 < length) {
-                // Convert four hex digits to codepoint. If any digit is invalid then the
-                // result is negative.
-                int code = (Character.digit(buffer[index++], 16) << 12) |
-                           (Character.digit(buffer[index++], 16) << 8) |
-                           (Character.digit(buffer[index++], 16) << 4) |
-                            Character.digit(buffer[index++], 16);
-
-                // If all digits are good.
-                if (code >= 0) {
-                    width = index - position;
-                    character = (char)code;
-
-                    return true;
-                }
-            }
+        // Needs to have been at least one u.
+        if (index == start) {
+            return false;
+        }
 
-            // Did not work out.
-            log.error(position, Errors.IllegalUnicodeEsc);
-            width = index - position;
+        int code = 0;
 
-            // Return true so that the invalid unicode escape is skipped.
-            return true;
+        for (int i = 0; i < 4; i++) {
+            int digit = Character.digit(buffer[index], 16);
+            code = code << 4 | digit;
+
+            if (code < 0) {
+                break;
+            }
+
+            index++;
         }
 
-        // Must be just a backslash.
-        character = '\\';
-        width = 1;
+        // Skip digits even if error.
+        width = index - position;
 
-        return false;
+        // If all digits are good.
+        if (code >= 0) {
+            character = (char)code;
+        } else {
+            log.error(position, Errors.IllegalUnicodeEsc);
+        }
+
+        // Return true even if error so that the invalid unicode escape is skipped.
+        return true;
     }
 
     /**
@@ -549,7 +550,7 @@ public class UnicodeReader {
         /**
          * Offset from the beginning of the original reader buffer.
          */
-        private int offset;
+        final private int offset;
 
         /**
          * Current column in the comment.
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8254073](https://bugs.openjdk.java.net/browse/JDK-8254073): Tokenizer improvements (revised)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to db7a2062fe4293b2b795ec2f5515180f9e28e449

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/525/head:pull/525`
`$ git checkout pull/525`

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java [hunk 1]
```diff
@@ -25,267 +25,500 @@
...
+
+        // Needs to have been at least one u.
+        if (index == start) {
+            return false;
+        }
+
+        int code = 0;
+
+        for (int i = 0; i < 4; i++) {
+            int digit = Character.digit(buffer[index], 16);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java:245]: This looks suspicious - what if index ends up being bigger than (or equal to)  `buffer.length` ?
Maybe we need a test for incomplete unicode sequences at the end of the tokenizer input - e.g. `\u123`

Comment (JimLaskey) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java:245]: You are correct. Will revise.

Comment (JimLaskey) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java]: Added change to 

    int digit = index < length ? Character.digit(buffer[index], 16) : -1;

Also added new test.

File: test/langtools/tools/javac/lexer/JavaLexerTest2.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import com.sun.tools.javac.parser.Tokens.Token;
+import com.sun.tools.javac.parser.Tokens.TokenKind;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.Log;
+
+import static com.sun.tools.javac.parser.Tokens.TokenKind.*;
+
+public class JavaLexerTest2 {
+    static final TestTuple[] TESTS = {
+            new TestTuple("0bL",         LONGLITERAL, true),
```
Comment (mcimadamore) [inline test/langtools/tools/javac/lexer/JavaLexerTest2.java:50]: Stylistic (optional) comment - we could have a common TestTuple superclass and two subclasses called Success and Failure, so that, by looking at the TESTS array it will be apparent what the behavior should be

File: test/langtools/tools/javac/lexer/JavaLexerTest2.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            new TestTuple("abc\\u0005def",IDENTIFIER, false),
+    };
+
+    static class TestTuple {
+        String input;
+        TokenKind kind;
+        String expected;
+        boolean willFail;
+
+        TestTuple(String input, TokenKind kind, String expected, boolean willFail) {
```
Comment (mcimadamore) [inline test/langtools/tools/javac/lexer/JavaLexerTest2.java:110]: Is anything using this?

File: test/langtools/tools/javac/lexer/JavaLexerTest2.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+import com.sun.tools.javac.parser.JavaTokenizer;
+import com.sun.tools.javac.parser.ScannerFactory;
+import com.sun.tools.javac.parser.Tokens.Token;
+import com.sun.tools.javac.parser.Tokens.TokenKind;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.Log;
+
+import static com.sun.tools.javac.parser.Tokens.TokenKind.*;
+
+public class JavaLexerTest2 {
```
Comment (mcimadamore) [inline test/langtools/tools/javac/lexer/JavaLexerTest2.java:48]: Should we merge JavaLexerTest and this one? After all you have all the required infra in here?

**Review by mcimadamore — APPROVED** [review]
Looks good - I've added optional suggestions for the test

File: test/langtools/tools/javac/lexer/JavaLexerTest2.java (hunk 2 above)
Comment (JimLaskey) [reply test/langtools/tools/javac/lexer/JavaLexerTest2.java]: Done

File: test/langtools/tools/javac/lexer/JavaLexerTest2.java (hunk 4 above)
Comment (JimLaskey) [reply test/langtools/tools/javac/lexer/JavaLexerTest2.java]: Done.

**Review by mcimadamore — APPROVED** [review]
Looks good
